### PR TITLE
Service accounts: add service account to team via db

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What this PR does / why we need it**:
This PR makes it possible for service accounts to be added to teams from the database side. I have purposefully made this PR very small to discuss subtleties in db implementation such as:

- do we want to be able to add service account functionality in the AddTeamMember sqlStore or should we create a separate implementation in the serviceaccounts.Store package, my quest would be yes. 🤔 
- need some input on the permissions parameter that we pass to `AddTeamMember()`

**Which issue(s) this PR fixes**:
https://github.com/grafana/grafana/issues/50375

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/serviceaccounts/database/database.go [hunk 1]
```diff
@@ -415,6 +415,11 @@ func (s *ServiceAccountsStoreImpl) SearchOrgServiceAccounts(
 	return searchResult, nil
 }
 
+// AddServiceAccountToTeam adds a service account to a team
+func (s *ServiceAccountsStoreImpl) AddServiceAccountToTeam(ctx context.Context, serviceAccountID int64, cmd *serviceaccounts.AddServiceAccountToTeamCommand) error {
+	return s.sqlStore.AddTeamMember(serviceAccountID, cmd.OrgId, cmd.TeamId, false, models.PERMISSION_ADMIN)
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/database/database.go:420]:
```suggestion
	return s.sqlStore.AddTeamMember(serviceAccountID, cmd.OrgId, cmd.TeamId, false, models.PERMISSION_ADMIN)
```

How are permissions passed down to this call?

---

Also; should we instead have a implementation of adding a service account to a team outside of `sqlStore` and make it part of the `serviceaccounts.Store` package instead? 🤔

File: pkg/services/serviceaccounts/models.go [hunk 2]
```diff
@@ -19,10 +19,6 @@ const (
 	ActionDelete = "serviceaccounts:delete"
 )
 
-type ServiceAccount struct {
-	Id int64
-}
-
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/models.go:25]: this was unused for some reason

File: pkg/services/serviceaccounts/database/database.go (hunk 1 above)
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/database/database.go:420]: should we instead have an implementation of adding a service account to a team outside of `sqlStore` and make it part of the `serviceaccounts.Store` package instead? 🤔

Comment (alexanderzobnin) [reply pkg/services/serviceaccounts/database/database.go:420]: We should specify with what permissions we add team member. I guess by default (if we don't want to let SA control team) it should be Viewer.

Comment (Jguer) [reply pkg/services/serviceaccounts/database/database.go:420]: I think it's good as it is. It has an interface, but we know for now the implementation is the same for users & service accounts.

Comment (eleijonmarck) [inline pkg/services/serviceaccounts/database/database.go:420]:
```suggestion
	return s.sqlStore.AddTeamMember(serviceAccountID, cmd.OrgId, cmd.TeamId, false, models.PERMISSION_VIEW)
```

File: pkg/services/serviceaccounts/database/database_test.go [hunk 3]
```diff
@@ -106,3 +106,35 @@ func TestStore_RetrieveServiceAccount(t *testing.T) {
...
+			sa := tests.SetupUserServiceAccount(t, db, c.sa)
+			team, err := db.CreateTeam(c.team.Name, c.team.Email, sa.OrgId)
+			if err != nil {
+				require.NoError(t, err)
+			}
+			cmd := &serviceaccounts.AddServiceAccountToTeamCommand{OrgId: sa.OrgId, TeamId: team.Id}
+			err = store.AddServiceAccountToTeam(context.Background(), sa.Id, cmd)
+			if c.expectedErr != nil {
+				require.ErrorIs(t, err, c.expectedErr)
+			}
```
Comment (Jguer) [inline pkg/services/serviceaccounts/database/database_test.go:137]: :triangular_ruler: Missing test of effect. You should be able to query the `team_member` table here to verify the user was indeed added with the correct permissions and (optionally) no other team member was modified

**Review by Jguer — CHANGES_REQUESTED** [review]

**Review by IevaVasiljeva — COMMENTED** [review]
When we set user membership for teams, we now use [team permission service](https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go#L36). We should explore how much work it would be to use the same flow for adding service accounts to teams. It might even already be possible through the current implementation, I'm not sure if/where we filter service accounts out.

However, even if it is already possible, I think we should do things properly and extend the [resource permission services](https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/resourcepermissions/service.go) to allow setting team permissions for service accounts. It would also make it easy to add dashboard, folder and data source permissions for service accounts if we ever decide to do that. 

Do you have any thoughts on this @kalleep?

**Comment by kalleep** [comment]
> When we set user membership for teams, we now use [team permission service](https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go#L36). We should explore how much work it would be to use the same flow for adding service accounts to teams. It might even already be possible through the current implementation, I'm not sure if/where we filter service accounts out.
> 
> However, even if it is already possible, I think we should do things properly and extend the [resource permission services](https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/resourcepermissions/service.go) to allow setting team permissions for service accounts. It would also make it easy to add dashboard, folder and data source permissions for service accounts if we ever decide to do that.
> 
> Do you have any thoughts on this @kalleep?

Yes I agree and I think using service account on the Team Permissions Service should work without any modifications

**Comment by Jguer** [comment]
I feel for this scenario (adding service accounts to teams), 

SAs should be completely treated as users to avoid any security issues from implementation and adding edge cases to team membership management

Even for the team UI I would:
Rename "users" to "users/service accounts" in teams UI
allow them to appear in the search for users in this case and just treat them as users for the addition

**Comment by eleijonmarck** [comment]
Thanks for your reviews!

To reiterate your thoughts ;

We would use team permission services w. The notion that SAs (service accounts) would simply be considered as users, i.e. not extending the service for now, but use it as users

**Comment by IevaVasiljeva** [comment]
> We would use team permission services w. The notion that SAs (service accounts) would simply be considered as users, i.e. not extending the service for now, but use it as users

Yes, let's try that and see how it works and whether it presents any issues.

File: pkg/services/serviceaccounts/database/database_test.go [hunk 4]
```diff
@@ -110,6 +111,69 @@ func TestStore_RetrieveServiceAccount(t *testing.T) {
...
+					sa.OrgId: {
+						accesscontrol.ActionTeamsRead:    []string{accesscontrol.ScopeTeamsAll},
+						accesscontrol.ActionOrgUsersRead: []string{accesscontrol.ScopeUsersAll},
+					},
+				},
+			}
+			teamQuery := models.GetTeamMembersQuery{OrgId: sa.OrgId, TeamId: team.Id, SignedInUser: signedInUser}
+			err = db.GetTeamMembers(context.Background(), &teamQuery)
+			require.NoError(t, err)
+			require.Equal(t, c.expectedNumberOfTeamMembers, len(teamQuery.Result))
```
Comment (eleijonmarck) [inline pkg/services/serviceaccounts/database/database_test.go:170]: @Jguer running the tests yields

```bash
go test -count 1 -v ./pkg/services/serviceaccounts/database -run TestStore_AddServiceAccountToTeam
```

=== RUN   TestStore_AddServiceAccountToTeam/should_be_able_to_add_service_account_to_team
    database_test.go:170:
        	Error Trace:	database_test.go:170
        	Error:      	Not equal:
        	            	expected: 1
        	            	actual  : 0

**Comment by eleijonmarck** [comment]
We will disregard this PR in favor of using the /api/teams route instead

**Pull request closed without merging**